x/gov/handler.go

```diff
@@ -67,17 +70,21 @@ func (h VoteHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (we
 
 func (h VoteHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
 	var res weave.DeliverResult
-	vote, proposal, elector, err := h.validate(ctx, db, tx)
+	voteMsg, proposal, vote, err := h.validate(ctx, db, tx)
 	if err != nil {
 		return res, err
 	}
-	if err := proposal.Vote(vote.Selected, *elector); err != nil {
+
+	if err := proposal.CountVote(*vote); err != nil {
 		return res, err
 	}
-	return res, h.propBucket.Update(db, vote.ProposalID, proposal)
+	if err = h.voteBucket.Save(db, h.voteBucket.Build(db, voteMsg.ProposalID, *vote)); err != nil {
+		return res, errors.Wrap(err, "failed to store vote")
+	}
+	return res, h.propBucket.Update(db, voteMsg.ProposalID, proposal)
 }
 
-func (h VoteHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*VoteMsg, *TextProposal, *Elector, error) {
+func (h VoteHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*VoteMsg, *TextProposal, *Vote, error) {
```

Comment:
non-blocking: life is hard when we return so many arguments, why don't we wrap all the structs in a struct?

Reply:
Agree that this is a smell. 3 + error is my personal limit to tolerate. Both `Check` and `Deliver` methods are super small so in this special case the caller would not benefit much from a new struct type. But good point and I will keep it in my head for the future.